devito/ops/compiler.py

```diff
@@ -30,3 +53,130 @@ def prepare_ops(self, soname, ccode, hcode):
+            warnings.simplefilter('ignore')
+
+            # Spinlock in case of MPI
+            sleep_delay = 0 if configuration['mpi'] else 1
+            _, _, cuda_o, _ = compile_from_string(
+                cuda_device_compiler, cuda_target,
+                cuda_code, cuda_src,
+                cache_dir=self.cache_dir,
+                debug=configuration['debug-compiler'],
+                sleep_delay=sleep_delay,
+                object=True
+            )
+            _, _, src_o, _ = compile_from_string(
+                cuda_host_compiler, self.target,
+                self.code, self.ops_src,
+                cache_dir=self.cache_dir,
+                debug=configuration['debug-compiler'],
+                sleep_delay=sleep_delay,
+                object=True
+            )
+            cuda_host_compiler.link_extension(
+                '%s%s' % (self.target, cuda_host_compiler.so_ext),
+                [src_o, cuda_o],
+                debug=configuration['debug-compiler']
+            )
+
+    def jit_compile(self, soname):
+        from devito import clear_cache
+        clear_cache()
+
+        self.target = str(self.get_jit_dir().joinpath(soname))
+        self.ops_src = '%s/%s_ops.cpp' % (self.get_jit_dir(), soname)
+        self.cache_dir = self.get_codepy_dir().joinpath(soname[:7])
+
+        # Typically we end up here
+        # Make a suite of cache directories based on the soname
+        self.cache_dir.mkdir(parents=True, exist_ok=True)
+        try:
+            with open(self.ops_src, 'r') as f:
+                self.code = f.read()
+        except FileNotFoundError:
+            warning("The file %s isn't present" % self.ops_src)
+        else:
+            self.cuda_compiler(soname)
+
+
+class CUDADeviceCompiler(CompilerOPS):
+    CXX = os.environ.get('CC', 'nvcc')
+
+    def __init__(self, *args, **kwargs):
+        super(CUDADeviceCompiler, self).__init__(*args, **kwargs)
+        self.o_ext = '.o'
+        self.cflags = ['-Xcompiler="-fPIC"', '-O3', '-g', '-gencode']
+        nv_arch = os.environ.get('NV_ARCH')
+
+        if(nv_arch is None):
+            raise ValueError("select an NVIDIA device to compile in CUDA, e.g. \
+                NV_ARCH=Kepler")
+        elif(nv_arch == 'Fermi'):
+            self.cflags.append('arch=compute_20,code=sm_21')
+        elif(nv_arch == 'Kepler'):
+            self.cflags.append('arch=compute_35,code=sm_35')
+        elif(nv_arch == 'Pascal'):
+            self.cflags.append('arch=compute_60,code=sm_60')
+        elif(nv_arch == 'Volta'):
+            self.cflags.append('arch=compute_70,code=sm_70')
+        else:
+            raise ValueError("Unknown NVIDIA architecture, select: \
+                Fermi, Kepler, Pascal or Volta")
+
+        self.ldflags = []
+
+        include_dirs = '%s %s/c/include' % (self.get_jit_dir(), self._ops_install_path)
+        self.include_dirs = include_dirs.split(' ')
+
+    def __lookup_cmds__(self):
```

Comment:
to be removed, this overwrites the `environ.get`

Reply:
Just removing causes an error.
I'll move `environ.get` to `__lookup_cmds__`.

Reply:
@mloubout do you agree with that?